File: agents/java/Agent.java
/**
 * @class Agent
 * @purpose Class that represents an agent which will play qchess
 */

import java.util.Random;
import java.util.Vector;

class Agent
{
	public Agent(String colour) throws Exception
	{
		this.colour = Piece.str2colour(colour);

		this.board = new Board();
		this.selected = null;
		this.rand = new Random();
	}

	public static void main(String [] args)
	{
		String colour = Reader.readLine();
		
		try
		{
			Agent agent = new Agent(colour);
			agent.Run();
		}
		catch (Exception e)
		{
			//System.err.println("EXCEPTION: agent.Run - "+e); 
			e.printStackTrace();
		}
	}

	public void Run() throws Exception
	{
		while (true)
		{
			Vector<String> v = Reader.readTokens();
			
			String cmd = v.get(0);
			
			if (cmd.compareTo("QUIT") == 0)
				break;
			else if (cmd.compareTo("SELECTION?") == 0)
			{
				Square s = select();
				System.out.println(Integer.toString(s.x) + " " + Integer.toString(s.y));
			}
			else if (cmd.compareTo("MOVE?") == 0)
			{
				Square s = move();
				System.out.println(Integer.toString(s.x) + " " + Integer.toString(s.y));
			}
			else
			{
				int x = Integer.parseInt(v.get(0));
				int y = Integer.parseInt(v.get(1));
				if (v.get(2).compareTo("->") == 0)
				{
					int x2 = Integer.parseInt(v.get(3));
					int y2 = Integer.parseInt(v.get(4));
					board.Update_move(x, y, x2, y2);
				}
				else
				{
					int index = Integer.parseInt(v.get(2));
					String type = v.get(3);
					board.Update_select(x, y, index, type);
				}
			}
		}
	}
		
	public Square select() throws Exception
	{
		Vector<Piece> p = board.pieces(colour);
		int choice = rand.nextInt(p.size());
		Square s = board.Get_square(p.get(choice).x, p.get(choice).y);
		if (s.piece == null)
			throw new Exception("ARGH");
		selected = s.piece;
		return s;
	}

	public Square move()
	{
		Vector<Square> v = new Vector<Square>();
		board.Get_moves(selected, v);
		return v.get(rand.nextInt(v.size()));
	}
	
	private final Piece.Colour colour; // colour of the agent; do not change it
	private Board board;
	private Piece selected; 
	private Random rand;
		
};

File: agents/silverfish/agent.cpp
/**
 * agent++ : A Sample agent for UCC::Progcomp2013
 * @file agent.cpp
 * @purpose Definition of Agent class
 */

#include "agent.h"
#include <cassert> // for sanity checks

using namespace std;

/**
 * @constructor Agent
 * @param new_colour - colour of the Agent
 */
Agent::Agent(const string & new_colour) : colour(Piece::str2colour(new_colour)), board(), selected(NULL)
{

}

/**
 * @destructor ~Agent
 */
Agent::~Agent()
{

}

/**
 * @funct Select
 * @purpose Selects a piece at random
 * @returns Square containing the selected piece
 */
Square & Agent::Select()
{
	vector<Piece*> & v = board.pieces(colour); // get pieces
	int choice = rand() % v.size(); // pick random index
	Piece * p = v[choice]; // get piece at the index
	assert(p->colour == colour);
	selected = p; // update selected
	//cerr << "Selected " << p->x << "," << p->y << " [" << p->types[0] << "," << p->types[1] << "]\n";
	return board.square(p->x, p->y); // get Square from board
}

/**
 * @funct Move
 * @purpose Pick a square to move a selected piece into
 * @returns Square to move last selected piece into
 */
Square & Agent::Move()
{
	assert(selected != NULL);
	vector<Square*> moves; // all possible moves for selected piece
	board.Get_moves(selected, moves); // populate possible moves
	assert(moves.size() > 0); 
	int choice = rand() % moves.size(); // pick random index
	return *(moves[choice]); // return that move
}

/**
 * @funct Run
 * @purpose The "Game Loop" for the agent; read commands and call appropriate function to make responses
 * @param in - Stream to read input from (use std::cin)
 * @param out - Stream to write output to (use std::cout)
 */
void Agent::Run(istream & in, ostream & out)
{
	string cmd; // buffer for tokens
	while (in.good())
	{
		in >> cmd; // read first token only
		if (cmd == "QUIT")
		{
			break;
		}
		else if (cmd == "SELECTION?")
		{
			Square & s = Select(); // get selection
			out << s.x << " " << s.y  << "\n"; // return response through output
		}
		else if (cmd == "MOVE?")
		{
			Square & s = Move(); // get move
			out << s.x << " " << s.y << "\n"; // return response through output
		}
		else
		{
			// There were multiple tokens...
			stringstream s(cmd);
			int x; int y;
			s >> x; // Convert first token (in cmd) to an int
			in >> y; // Read second token from in
			
			in >> cmd; // Read third token
			
			if (cmd == "->") // Token indicates a move was made
			{
				int x2; int y2; // remaining two tokens indicate destination
				in >> x2; in >> y2;
				board.Update_move(x, y, x2, y2); // update the board
			}
			else
			{
				// Tokens are for a selection
				int index; stringstream s2(cmd); 
				s2 >> index; // convert third token to an index
				in >> cmd; // Read fourth token - the new type of the piece
				board.Update_select(x, y, index, cmd); // update the board
			}
			
		}
	}
}
